[ops] Dashboard MCP calls fail: Unauthorized — bearer token resolved to codex

## 증상

Dashboard agent의 MCP tool 호출(`masc_persona_list`, `masc_keeper_status`)이 반복적으로 401 Unauthorized로 실패합니다.

에러 메시지: `No credential found for dashboard (bearer token belongs to codex)`

## 재현 조건

1. 서버 실행 중 (`main_eio.exe --base-path=/Users/dancer/me`)
2. Dashboard agent가 `masc_persona_list` 또는 `masc_keeper_status` 호출
3. 2026-04-23 09:49~18:50 사이 지속적으로 발생

## 로그

```
[ERROR] [MCP] tool call failed: masc_persona_list — detail=🔐 Unauthorized: No credential found for dashboard (bearer token belongs to codex)
[ERROR] [MCP] tool call failed: masc_keeper_status — detail=🔐 Unauthorized: No credential found for dashboard (bearer token belongs to codex)
```

## 추정 원인

- `dashboard` agent에 대한 별도 credential(`dashboard.json`)이 `.masc/auth/agents/`에 없을 가능성
- 클라이언트가 codex의 raw token을 dashboard 호출에 사용 중일 가능성
- `find_credential_by_token`에서 token hash가 codex의 것과 매칭됨

## 관련

- 다른 반복 이상 상태: #9620 (sangsu token mismatch), #9629 (Governance timeout), #9572 (nick0cave cascade_name)

--- a/dashboard/src/api/mcp.test.ts
+++ b/dashboard/src/api/mcp.test.ts
@@ -55,6 +55,7 @@ beforeEach(() => {
   // mocks whose `vi.hoisted` defaults have been overwritten by an earlier
   // test (e.g. authHeaders). The dev-token bootstrap must stay inert unless
   // a test explicitly exercises it.
+  currentDashboardActor.mockReturnValue('dashboard')
   getStoredToken.mockReturnValue('test-stored-token')
   getStoredTokenMeta.mockReturnValue({
     source: 'manual',
@@ -168,6 +169,66 @@ describe('dev-token bootstrap', () => {
     expect(calls.length).toBeGreaterThanOrEqual(2)
     expect(calls[0]?.[0]).toBe('/api/v1/dashboard/dev-token')
     expect(calls[1]?.[0]).toBe('/mcp')
+  }, 60_000)
+
+  it('replaces a borrowed loopback token when the default dashboard actor is active', async () => {
+    getStoredToken.mockReturnValue('borrowed-codex-token')
+    getStoredTokenMeta.mockReturnValue({
+      source: 'manual',
+      actor: null,
+      scope: null,
+    })
+    fetchWithTimeout
+      .mockResolvedValueOnce(
+        new Response(JSON.stringify({
+          token: 'loopback-dev-token',
+          actor: 'dashboard',
+          scope: 'admin',
+        }), {
+          status: 200,
+          headers: { 'Content-Type': 'application/json' },
+        }),
+      )
+      .mockResolvedValueOnce(
+        new Response('{}', { status: 200, headers: { 'Mcp-Session-Id': 'sess-borrowed' } }),
+      )
+      .mockResolvedValueOnce(new Response('', { status: 202 }))
+      .mockResolvedValueOnce(
+        new Response('data: {"result":{"content":[{"type":"text","text":"ok"}]}}\n', { status: 200 }),
+      )
+
+    const { callMcpTool } = await import('./mcp')
+    await callMcpTool('masc_status', {})
+
+    expect(setStoredToken).toHaveBeenCalledWith('loopback-dev-token', {
+      source: 'dev',
+      actor: 'dashboard',
+      scope: 'admin',
+    })
+    const calls = fetchWithTimeout.mock.calls as Array<[string, RequestInit]>
+    expect(calls[0]?.[0]).toBe('/api/v1/dashboard/dev-token')
+  }, 60_000)
+
+  it('keeps manual tokens for non-default dashboard actors', async () => {
+    currentDashboardActor.mockReturnValue('dashboard-manual-actor')
+    authHeaders.mockReturnValue({
+      'Authorization': 'Bearer manual-actor-token',
+      'X-MASC-Agent': 'dashboard-manual-actor',
+    })
+    getStoredToken.mockReturnValue('manual-actor-token')
+    getStoredTokenMeta.mockReturnValue({
+      source: 'manual',
+      actor: null,
+      scope: null,
+    })
+    setupMcpSessionMocks('sess-manual-actor')
+
+    const { callMcpTool } = await import('./mcp')
+    await callMcpTool('masc_status', {})
+
+    expect(setStoredToken).not.toHaveBeenCalled()
+    const calls = fetchWithTimeout.mock.calls as Array<[string, RequestInit]>
+    expect(calls[0]?.[0]).toBe('/mcp')
   }, 60_000)
 
   it('swallows dev-token fetch failures so strict-auth servers still reach the 401 path', async () => {

--- a/dashboard/src/api/mcp.ts
+++ b/dashboard/src/api/mcp.ts
@@ -42,10 +42,11 @@ function shouldRefreshDevToken(): boolean {
   const meta = getStoredTokenMeta()
   if (!token) return true
   if (meta?.source === 'dev') return true
-  // Legacy sessions only stored the raw token string. On loopback with the
-  // default dashboard actor, treat them as managed-token candidates so
-  // frequent restarts or base_path flips self-heal without manual clearing.
-  return meta == null && !isRemoteAccess() && currentDashboardActor() === 'dashboard'
+  const actor = currentDashboardActor()
+  if (isRemoteAccess() || actor !== 'dashboard') return false
+  // Loopback dashboard sessions should self-heal if they are still holding
+  // a borrowed non-dashboard token (for example an old codex paste/URL token).
+  return meta == null || meta.actor == null || meta.actor !== actor
 }
 
 /** Fetch the loopback-only dev token once per page load and stash it so

